src/main/scala/chisel3/util/LoadMemoryTransform.scala

```diff
@@ -0,0 +1,234 @@
+      )
+    case _ =>
+  }
+
+  def transformClass : Class[LoadMemoryTransform] = classOf[LoadMemoryTransform]
+
+  def toFirrtl: LoadMemoryAnnotation = LoadMemoryAnnotation(target.toNamed, fileName, hexOrBinary)
+}
+
+/**
+  * Firrtl implementation for load memory
+  * @param target        memory to load
+  * @param fileName      name of input file
+  * @param hexOrBinary   use $readmemh or $readmemb
+  */
+case class LoadMemoryAnnotation(
+  target: Named,
+  fileName: String,
+  hexOrBinary: String = "h",
+  originalMemoryNameOpt: Option[String] = None
+) extends SingleTargetAnnotation[Named] {
+
+  def getFileName: String = {
+    (target, originalMemoryNameOpt) match {
+      case (ComponentName(componentName, _), Some(originalMemoryName)) =>
+        if(componentName == originalMemoryName) {
+          fileName
+        }
+        else {
+          fileName + componentName.drop(originalMemoryName.length)
+        }
+      case _ =>
+        fileName
+    }
+  }
+
+  def duplicate(n: Named): LoadMemoryAnnotation = {
+//    if(n.serialize != target.serialize) {
+////      Builder.errors.error(s"LoadMemory[${target.serialize}] must be a simple memory, no bundles")
+//      throw new Exception(s"LoadMemory[${target.serialize}] must be a simple memory, no bundles")
+//    }
+    target match {
+      case ComponentName(componentName, _) =>
+        this.copy(target = n, originalMemoryNameOpt = Some(componentName))
+      case _ =>
+        this.copy(target = n)
+    }
+  }
+}
+
+/**
+  * This pass creates BlackBoxInlineAnno from the LoadMemoryAnnotations
+  * it does this even if the backend is not verilog.
+  *
+  * @param circuitState the target circuit state
+  */
+//TODO: (chick) support a treadle or interpreter means of memory loading
+//TODO: (chick) can this only be done when backend is known to support this.
+//scalastyle:off method.length cyclomatic.complexity regex
+class CreateBindableMemoryLoaders(circuitState: CircuitState) extends Pass {
+
+  val annotations      : Seq[Annotation] = circuitState.annotations
+  val memoryAnnotations: Seq[LoadMemoryAnnotation] = annotations.collect{ case m: LoadMemoryAnnotation => m }
+
+  val bindModules      : mutable.ArrayBuffer[BlackBoxInlineAnno] = new mutable.ArrayBuffer()
+
+  val verilogEmitter   : VerilogEmitter = new VerilogEmitter
+
+  /**
+    * finds the specified module name in the circuit
+    *
+    * @param moduleName name to find
+    * @param circuit circuit being analyzed
+    * @return the circuit, exception occurs in not found
+    */
+  def findModule(moduleName: String, circuit: Circuit): DefModule = {
+    circuit.modules.find(module => module.name == moduleName) match {
+      case Some(module: firrtl.ir.Module) =>
+        module
+      case Some(externalModule: DefModule) =>
+        externalModule
+      case _ =>
+        throw new Exception(s"Could not find module $moduleName in circuit $circuit")
```

Comment:
Using the error system between firrtl and chisel has been a problem for me. Don't tell anyone but half the reason I implemented the composite memory logic was that I was having trouble figuring out how to generate errors when that occurred.  
@jackkoenig what's the best way to do this. This transform is in chisel3 but is invoked in firrtl. All too often errors I generate create the particularly unhelpful message "File a Firrtl Issue" even though in my case it's a simple user error.  Should I follow the model of something like `CheckComboLoops` or is there a better example. I looked around a little. This seems like it would be pretty useful wiki page

Reply:
No reason for this to hold it up either. My (unofficial) strategy has been to have anything run by the tool throw that type of exception. It's entirely unnecessary, though. Yeah, no need for `throwInternalError` (we don't want new issues filed here...). Maybe a question for the group and to be added to the wiki as you suggest.